vfio/src/vfio_pci.rs

```diff
@@ -676,12 +693,18 @@ impl VfioPciDevice {
 
 impl Drop for VfioPciDevice {
     fn drop(&mut self) {
-        if self.interrupt.msi.is_some() && self.device.disable_msi().is_err() {
-            error!("Could not disable MSI");
-        }
-
-        if self.interrupt.msix.is_some() && self.device.disable_msix().is_err() {
-            error!("Could not disable MSI-X");
```

Comment:
You don't need to introduce `InterruptType` and store it to prevent from calling the disabling function twice.
All you need is to check if it is still enabled for both MSI and MSI-X.
```
impl Drop for VfioPciDevice {
    fn drop(&mut self) {
        if let Some(msi) &self.interrupt.msi {
            if msi.enabled() && self.device.disable_msi().is_err() {
                error!("Could not disable MSI");
            }
        }

        if let Some(msix) = &self.interrupt.msix {
            if msix.enabled() && self.device.disable_msix().is_err() {
                error!("Could not disable MSI-X");
            }
        }

        if self.device.unset_dma_map().is_err() {
            error!("failed to remove all guest memory regions from iommu table");
        }
    }
}
```

Let me know if this simple patch work for you.

Reply:
You're right, I will modify it and push a new patch.